README/RCodingStyle.hxx

```diff
@@ -0,0 +1,171 @@
+   /// This is a nested class. Only use `struct`s for simple aggregates without
+   /// methods and all public members.
+   struct Nested {
+      std::string fNestedName; ///< The nested name.
+   };
+
+private:
+   // SECOND section: data members.
+
+   int fIntMember = -2; ///< Use inline initialization at least for fundamental types.
+   std::vector<int> fManyInts; ///< This documents a regular member. Don't explicitly default initialize (`{}`)
+   std::vector<Nested> fManyStrings; ///<! This marks a member that does not get serialized.
+   static constexpr const SeqULL_t kSeq; ///< A static constexpr const variable starts with `k`.
+
+   /// A static variable starts with `fg`. Avoid all non-constexpr static variables! Long data
+   /// member documentation is just fine, but then put it in front of the data member.
+   static std::string fgName;
+
+   /// THIRD section: non-public functions.
+
+   /// It's a good habit to prevent slicing of classes that can be base classes,
+   /// e.g. for `RExampleClass &Get()` you likely want to keep the derived object intact,
+   /// and prevent `auto ret = Get()` (where `ret` is now an object of type `RExampleClass`).
+   RExampleClass(const RExampleClass&) = default;
+   RExampleClass(RExampleClass&&) = default;
+
+   /// Instead of static data members (whether private or public), use outlined static
+   /// functions. See the implementation in the source file.
+   static const std::string &StaticFunc();
+
+public:
+   // THIRD section: methods, starting with constructors, then alphabetical order.
+
+   /// Use `= default` inside the class whenever possible
+   RExampleClass() = default;
+
+   /// Classes with virtual functions must have a virtual destructor. If all other functions are
+   /// pure virtual or inlined, the destructor should be outlined to pin the class's vtable to an
+   /// object file. I.e. say `= default` in the source, not here.
+   virtual ~RExampleClass();
+
+   /// Virtual functions are still fine (though we try to avoid them as they come with a call
+   /// and optimization penalty). See `RCodingStyle` for overriding virtual functions.
+   virtual void AVirtualFunction() = 0;
+
+   /// Static functions line up alphabetically with the others.
+   static SeqULL_t GetSeq() { return fkSeq; }
+
+   /// "Free" (i.e. non-class member) operators should be declared as "hidden friends".
+   /// This makes them available only when needed, instead of adding to the long list of
+   /// "no matching overload for a+b, did you mean ...(hundreds of unrelated types)..."
+   friend RExampleClass operator+(const RExampleClass &a, const RExampleClass &b) {
+      return {};
+   }
+
+};
+
+///\class RCodingStyle
+/// This is exhaustive documentation of the class template, explaining the constraints
+/// on the template parameters, e.g. `IDX` must be positive, `T` must be  copyable.
+/// Disabled if `T` is of reference type.
+
+template <int IDX, // Template parameters are all-capital letters
```

Comment:
this is the only new convention that I don't like, it makes template-heavy code a bit heavy to look at (just go to `TypeTraits.hxx` and capitalize all template parameters...) and macro-like.

Reply:
It helps to identify template parameters as such. I often use a using declaration to "convert" the template param to a type (with type-spelling).